src/main/java/org/codeseoul/event_member_management/member/MemberController.java

```diff
@@ -33,49 +36,84 @@ public class MemberController {
         return assembler.toModel(member);
     }
 
     @PutMapping("/members/{id}")
-    EntityModel<Member> replaceMember(@RequestBody Member member, @PathVariable Long id) {
-
-        return repository.findById(id)
-                .map(dbMember -> {
-                    dbMember.setDisplayName(member.getDisplayName());
-                    dbMember.setEmail(member.getEmail());
-                    dbMember.setPhoneNumber(member.getPhoneNumber());
-                    dbMember.setImageUrl(member.getImageUrl());
-                    dbMember.setShortBio(member.getShortBio());
-                    return assembler.toModel(repository.save(dbMember));
-                })
-                .orElseGet(() -> {
-                    member.setId(id);
-                    return assembler.toModel(repository.save(member));
-                });
+    EntityModel<Member> replaceMember(@Valid @RequestBody Member newMember, @PathVariable Long id, BindingResult errors) throws MethodArgumentNotValidException {
+        Optional<Member> dbMember = repository.findById(id);
+        if (dbMember.isPresent()) {
+            Member oldMember = dbMember.get();
+            validateExistingMember(oldMember, newMember, errors);               
+            oldMember.setUsername(newMember.getUsername());
+            oldMember.setDisplayName(newMember.getDisplayName());
+            oldMember.setEmail(newMember.getEmail());
+            oldMember.setPhoneNumber(newMember.getPhoneNumber());
+            oldMember.setImageUrl(newMember.getImageUrl());
+            oldMember.setShortBio(newMember.getShortBio());
+            return assembler.toModel(repository.save(oldMember));
+        } else {
+            throw new MemberNotFoundException(id);
+        }
     }
 
     @DeleteMapping("/members/{id}")
     void deleteMember(@PathVariable Long id) {
+        repository.findById(id)
+            .orElseThrow(() -> new MemberNotFoundException(id));
         repository.deleteById(id);
     }
+
+    private void validateExistingMember(Member oldMember, Member newMember, BindingResult errors) throws MethodArgumentNotValidException {
+        if (!oldMember.getUsername().equals(newMember.getUsername()) && repository.findByUsername(newMember.getUsername()).isPresent())
+            errors.addError(new FieldError("Member", "username", "username already in use"));
+        if (!oldMember.getEmail().equals(newMember.getEmail()) && repository.findByEmail(newMember.getEmail()).isPresent())
+            errors.addError(new FieldError("Member", "email", "email already in use"));
+        if (!oldMember.getPhoneNumber().equals(newMember.getPhoneNumber()) && repository.findByPhoneNumber(newMember.getPhoneNumber()).isPresent())
+            errors.addError(new FieldError("Member", "phoneNumber", "phoneNumber already in use"));
+        if (errors.hasErrors())
+            throw new MethodArgumentNotValidException(null, errors);
```

Comment:
I think we can define a new repository method like `findByUsernameOrEmailOrPhoneNumber` to reduce the number of queries as you mentioned. JPA seems to support this: https://docs.spring.io/spring-data/jpa/docs/current/reference/html/#repositories.query-methods.query-creation

Very cool on the binding result. I didn't know about that. I actually don't know Spring that well, yet.

Reply:
Wow thanks, I didn't know you could be that explicit with the repository methods. I'll give it a go tonight.